sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,500 @@
+
+```c++
+int vectorSum(float *v, size_t s, size_t sz) {
+  int result = 0;
+  for (size_t i = s; i < s+sz; i++)
+    result += v[i];
+  return result;
+}
+
+using namespace sycl::ext::intel::experimental;
+
+{
+  sycl::queue q{...};
+
+  constexpr count = ...;
+  std::vector<float> in(count), out(2);
+  ...
+
+  sycl::buffer in_buf(in);
+  sycl::buffer out_buf(out);
+
+  q.submit([&](sycl::handler& h) {
+    sycl::accessor in_acc(in_buf, h, sycl::read_only);
+    sycl::accessor out_acc(out_buf, h, sycl::write_only);
+    h.single_task<ParallelTask>([=]() {
+      task_sequence<vectorSum> firstQuarter;
+      task_sequence<vectorSum> secondQuarter;
+      task_sequence<vectorSum> thirdQuarter;
+      task_sequence<vectorSum> fourthQuarter;
+      int quarterCount = count/4;
+      firstQuarter.async(in_acc.get_pointer(), 0, quarterCount);
+      secondQuarter.async(in_acc.get_pointer(), quarterCount, quarterCount);
+      thirdQuarter.async(in_acc.get_pointer(), 2*quarterCount, quarterCount);
+      fourthQuarter.async(in_acc.get_pointer(), 3*quarterCount, quarterCount);
+      out_acc[1] = firstQuarter.get() + secondQuarter.get() +
+        thirdQuarter.get() + fourthQuarter.get();
+    });
+  }).wait();
+  ...
+}
+```
+
+== Definitions
+
+== `task_sequence` class
+
+[source,c++,linenums]
+----
+namespace sycl::ext::intel::experimental {
+
+template<auto& f, class propertiesT =
+   decltype(oneapi::experimental::properties{})>
+class task_sequence{};
+
+template <typename ReturnT, typename... ArgsT, ReturnT (&f)(ArgsT...),
+          class propertiesT>
+class task_sequence<f, propertiesT> {
+
+public:
+  task_sequence(const task_sequence &) = delete;
+  task_sequence &operator=(const task_sequence &) = delete;
+  task_sequence(task_sequence &&) = delete;
+  task_sequence &operator=(task_sequence &&) = delete;
+
+  task_sequence();
+
+  task_sequence(const propertiesT &);
+
+  void async(ArgsT... args);
+
+  ReturnT get();
+  
+  template <typename propertyT>
+  static constexpr bool has_property();
+  
+  template <typename propertyT>
+  static constexpr auto get_property();
+
+  ~task_sequence();
+};
+
+} // namespace experimental
+} // namespace intel
+} // namespace ext
+} // namespace sycl
+
+----
+
+`task_sequence` is a class template, parameterized by an `auto` reference to a
++Callable+ `f` and by a list of properties `propertiesT`.
+The +Callable+ `f` defines the asynchronous task to be
+associated with the `task_sequence`, and requiring an auto reference ensures
+that each `f` be statically resolvable at compile time.  Static resolvability by
+the compiler is desirable when compiling for spatial architectures as it can
+enable the generation of more efficient hardware. Furthermore, the partial
+specialization `typename ReturnT, typename ... ArgsT, ReturnT(&f) (ArgsT...)`
+ensures that the return type `ReturnT` of `f` and the argument types
+`typename ... ArgsT` of `f` are resolvable and fixed for each definition of
```

Comment:
Isn't this already guaranteed by templating on an auto reference?  For example, if you didn't have the partial specialization and you tried to declare a task_sequence object templated on an overloaded function, you'd get a compiler error because the type of f couldn't be deduced (proof: https://godbolt.org/z/K8zo67Y5W).  It seems to be that the partial specialization is for our convenience to make writing the class easier and not to enforce an additional constraint on the caller.

Reply:
Updated to say that it makes it easier to write the class - although I think this should be an implementation detail and not mentioned in the spec at all. Do you agree?